src/LuaTransformer.ts

```diff
@@ -3003,18 +3012,28 @@ export class LuaTransformer {
         );
     }
 
-    public transformFunctionExpression(
-        node: ts.FunctionLikeDeclaration,
-        context: tstl.Identifier | undefined
-    ): ExpressionVisitResult
-    {
+    public transformFunctionExpression(node: ts.FunctionLikeDeclaration): ExpressionVisitResult {
         const type = this.checker.getTypeAtLocation(node);
-        const hasContext = tsHelper.getFunctionContextType(type, this.checker) !== ContextType.Void;
+
+        let context: tstl.Identifier | undefined;
+        if (tsHelper.getFunctionContextType(type, this.checker) !== ContextType.Void) {
+            if (ts.isArrowFunction(node)) {
+                // dummy context for arrow functions with parameters
+                if (node.parameters.length > 0) {
+                    context = tstl.createAnnonymousIdentifier();
+                }
+            } else {
+                // self context
+                context = this.createSelfIdentifier();
+            }
+        }
+
         // Build parameter string
-        const [paramNames, dotsLiteral, spreadIdentifier] = this.transformParameters(
-            node.parameters,
-            hasContext ? context : undefined
-        );
+        const [paramNames, dotsLiteral, spreadIdentifier] = this.transformParameters(node.parameters, context);
+
+        let flags = !ts.isFunctionDeclaration(node)
```

Comment:
This statement looks very strange, when is a functionExpression not a functionDeclaration/

Reply:
Perhaps this flag can be named better - it's a hint for the printer to output as an assignment expression `foo = function()` instead of a function declaration `function foo()`

Reply:
It's not the flag that I'm confused by, but the condition here.